### DISCUSSION TITLE
Recommended way to interpolate dynamical matrices?

### DISCUSSION THREAD

**Discussion by bkchang** [SSCHAcode/python-sscha#116] [Q&A] [answered]
Dear developers,

I know this question has been raised a couple of times, but I'd still want to make sure how to properly interpolate the resulting dynamical matrices obtained from SSCHA calculations, as I feel like you are still frequently updating your code, and the previous answers are not entirely clear to me. Below is my problem description:

Starting from the dynamical matrices (which result in soft modes) computed from DFPT using a 2x2x2 q-mesh in Quantum espresso, I was able to use your SSCHA code to obtain a set of new dynamical matrices defined on this exact same 2x2x2 q-mesh that gives in all-positive phonons. But I'd like to interpolate these dynamical matrices to a 4x4x4 q-mesh, and not just any 4x4x4 q-mesh, but the 4x4x4 irreducible q-mesh Quantum Espresso would automatically generate for this system if I ask it to.

I checked both your answers to the previous discussions as well as your user guide. In [a previous post](https://github.com/SSCHAcode/python-sscha/discussions/74#discussion-3971454) I saw people using `Phonon.Phonon.Interpolate` to do this kind of interpolation. But then I checked your CellConstructor user guide, which says "NOTE: This method ignores effective charges... This is going to be replaced with the InterpolateMesh function, accounting properly for effective charges". And so I gave a try to `Phonon.Phonon.InterpolateMesh`, which, according to the user guide, only requires one input variable, which is an array of dimensions of the supercell I'd like to interpolated to. So I did `Phonon.Phonon.InterpolateMesh([4, 4, 4])`, and I found that, although it generates the dynamical matrices on a correct number of q-points (equal to the number of irreducible q-points that QE would generate for 4x4x4 for my system), those newly generated q-points are still very different from QE's irreducible 4x4x4.

In summary, my question is:
What is the best / correct way to interpolate the SSCHA dynamical matrices to a denser QE irreducible grid while taking care of the effective charges?

Thank you very much for you patience and help in advance.

Best,
Ben

**Comment by mesonepigreco** [comment, marked as answer]
Hi Ben, 
I'm very sorry for the late reply; for some reason, this post was not notified correctly through the GitHub system and slipped my attention. I hope you found a solution, I will answer anyway as it could be of interest also for other people. 
Both Interpolate and InterpolateMesh chose the correct q-mesh, the same as quantum espresso.
There could be a difference in the order of q-points, or some q-points could be shifted by the reciprocal vectors, but they would be precisely the same.

You need to call the following function to group the q points in the stars as quantum espresso does:
```
dyn.AdjustQPoints()
```
After the Interpolate or InterpolateMesh, it should be done automatically.

However, the best way to interpolate the dynamical matrices also exploiting the effective charges is by using support harmonic dynamical matrices in the Interpolate method.

You can call Interpolating as 

```
# dyn_harm_222 is the harmonic dyn in a 2x2x2 q mesh
# dyn_harm_444 is the harmonic dyn in a 4x4x4 q mesh
# dyn_sscha_222 is the sscha dyn in the 2x2x2 q mesh

dyn_sscha_444 = dyn_sscha_222.Interpolate( (2,2,2), (4,4,4), dyn_harm_222, dyn_harm_444)
```
In this way, the q mesh will be the same as the one of dyn_harm_444 computed with quantum espresso, and it will exploit the two harmonic dynamical matrices to improve the interpolation (therefore, it accounts also for long-range interactions even better than using just the effective charges).

I hope that this clarify the problem, 
Bests,
Lorenzo
